controlpanel/api/auth0.py

```diff
@@ -0,0 +1,211 @@
+        )
+        self.domain = kwargs.get("domain", settings.AUTH0["domain"])
+        self._access_token = None
+
+    @property
+    def access_token(self):
+        if self._access_token is None:
+            get_token = authentication.GetToken(self.domain)
+
+            try:
+                token = get_token.client_credentials(
+                    self.client_id, self.client_secret, self.audience
+                )
+
+            except exceptions.Auth0Error as error:
+                raise Auth0Error(
+                    f"Failed getting Auth0 access token for client "
+                    f"{self.client_id} for audience {self.audience} "
+                    f"at {self.domain}: {error}"
+                )
+
+            self._access_token = token["access_token"]
+
+        return self._access_token
+
+    def request(self, method, endpoint, **kwargs):
+        base_url = self.base_url
+        if not base_url.endswith('/'):
+            base_url = base_url + '/'
+        url = f"{base_url}{endpoint}"
+        response = requests.request(
+            method,
+            url,
+            headers={
+                "Content-Type": "application/json",
+                "Authorization": f"Bearer {self.access_token}",
+            },
+            **kwargs,
+        )
+        response.raise_for_status()
+
+        if response.text:
+            return response.json()
+
+        return {}
+
+    def get_all(self, endpoint, key=None, **kwargs):
+        items = []
+        total = None
+        if "params" not in kwargs:
+            kwargs["params"] = {}
+        if "page" in kwargs:
+            kwargs["params"]["page"] = kwargs.pop("page")
+        if "per_page" in kwargs:
+            kwargs["params"]["per_page"] = kwargs.pop("per_page")
+
+        if key is None:
+            key = endpoint
+
+        while True:
+            response = self.request("GET", endpoint, **kwargs)
+
+            if "total" not in response:
+                raise Auth0Error(f"get_all {endpoint}: Missing total")
+
+            if key not in response:
+                raise Auth0Error(f"get_all {endpoint}: Missing key {key}")
+
+            if total is None:
+                total = response["total"]
+
+            if total != response["total"]:
+                raise Auth0Error(f"get_all {endpoint}: Total changed")
+
+            items.extend(response[key])
+
+            if len(items) >= total:
+                break
+
+            if len(response[key]) < 1:
+                break
+
+            if "page" not in kwargs["params"]:
+                kwargs["params"]["page"] = 1
+
+            kwargs["params"]["page"] += 1
+
+        return items
+
+
+class AuthorizationAPI(APIClient):
+    base_url = settings.AUTH0["authorization_extension_url"]
+    audience = "urn:auth0-authz-api"
+
+    def get_users(self):
+        return self.get_all("users", page=0, per_page=100)
+
+    def get_group(self, group_name):
+        groups = [
```

Comment:
Why do you need this list? Isn't `get_all("groups")` returning a list already? Couldn't you assign that to `groups` and the rest of the logic stays the same? Probably there is a good reason why you're using list comprehension here.

Reply:
Did you miss the `if group["name"] == group_name` in the list comprehension? This is filtering the list of groups from `get_all("groups")` to the one (or more if there are duplicates) named group. The Authorization API only allows you to get a single group by ID, not by name, so we have to fetch all the groups and filter like this.

Reply:
Makes sense, I probably missed something when I had a look 👍